lib/src/test/java/io/ably/lib/transport/HostsTest.java

```diff
@@ -0,0 +1,271 @@
+        String host = hosts.getFallback(Defaults.HOST_REALTIME);
+        assertThat(host, nullValue());
+    }
+
+    /**
+     * Expect that returned host is contained within custom host list in options,
+     * but in shuffled order and the right number of them.
+     */
+    @Test
+    public void hosts_fallback_custom_hosts() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        String[] customHosts = { "F.ably-realtime.com", "G.ably-realtime.com", "H.ably-realtime.com", "I.ably-realtime.com", "J.ably-realtime.com", "K.ably-realtime.com" };
+        options.fallbackHosts = customHosts;
+        Hosts hosts = new Hosts(Defaults.HOST_REALTIME, Defaults.HOST_REALTIME, options);
+        int idx;
+        String host = Defaults.HOST_REALTIME;
+        boolean shuffled = false;
+        boolean allFound = true;
+        for (idx = 0; ; ++idx) {
+            host = hosts.getFallback(host);
+            if (host == null)
+                break;
+            int found = Arrays.asList(customHosts).indexOf(host);
+            if (found < 0)
+                allFound = false;
+            else if (found != idx)
+                shuffled = true;
+        }
+        assertThat(idx, is(customHosts.length));
+        assertTrue(allFound);
+        assertTrue(shuffled);
+    }
+
+    /**
+     * Expect that returned host is contained within default host list
+     */
+    @Test
+    public void hosts_fallback_no_custom_hosts() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        Hosts hosts = new Hosts(null, Defaults.HOST_REALTIME, options);
+        int idx;
+        String host = Defaults.HOST_REALTIME;
+        boolean shuffled = false;
+        boolean allFound = true;
+        for (idx = 0; ; ++idx) {
+            host = hosts.getFallback(host);
+            if (host == null)
+                break;
+            int found = Arrays.asList(Defaults.HOST_FALLBACKS).indexOf(host);
+            if (found < 0)
+                allFound = false;
+            else if (found != idx)
+                shuffled = true;
+        }
+        assertThat(idx, is(Defaults.HOST_FALLBACKS.length));
+        assertTrue(allFound);
+        assertTrue(shuffled);
+    }
+
+    /**
+     * Expect a null, when realtimeHost is non-default
+     */
+    @Test
+    public void hosts_fallback_overridden_host() throws AblyException {
+        ClientOptions options = new ClientOptions();
+        String host = "overridden.ably.io";
+        Hosts hosts = new Hosts(host, Defaults.HOST_REALTIME, options);
+        host = hosts.getFallback(host);
+        assertThat(host, nullValue());
+    }
+
+    /**
+     * Expect that returned host is contained within default host list
+     * when realtimeHost is non-default and fallbackHostsUseDefault is set
+     */
+    @Test
+    public void hosts_fallback_use_default() throws AblyException {
```

Comment:
Can we make the test a bit simple. I think Array. Equals should do the job if it does a deep check. The test looks a bit complex to understand at first look.

Reply:
Or maybe take out common logic in test helpers. But, if abstraction already exists, then we should reuse it.

Reply:
There are was a bunch of tests which were there before and it seems that I wrote them, but I've just moved this file. Yes, the logic is more complicated than it should be, but I checked the tests and they look good to me. I would be glad to make these tests nicer but in a separate pull request.